### PR TITLE
PIP-45: Migrate OwnershipCache to use MetadataStore

### PR DESCRIPTION
**Pull request by merlimat**
### Motivation

Migrate OwnershipCache logic for acquiring and watching bundles ownership to use ResourceLocks from the MetadataStore interface.

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 1]
```diff
@@ -631,168 +631,146 @@ protected void handleProducerRemoved(Producer producer) {
                                                  Map<String, String> metadata, boolean readCompacted,
                                                  InitialPosition initialPosition,
                                                  long startMessageRollbackDurationSec,
-                                                 boolean replicatedSubscriptionState,
+                                                 boolean replicatedSubscriptionStateArg,
                                                  KeySharedMeta keySharedMeta) {
+        return brokerService.checkTopicNsOwnership(getName()).thenCompose(__ -> {
+            if (readCompacted && !(subType == SubType.Failover || subType == SubType.Exclusive)) {
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:637]: we could move this validation outside of the `checkTopicNsOwnership` block.
not a big deal, it is something that does not happen in the happy path

**Review by eolivelli — COMMENTED** [review]
I did a first pass.
the patch is very big and I am not used to this code but overall it looks good

**Comment by Anonymitaet** [comment]
@merlimat thanks for your contribution. For this PR, do we need to update docs?

**Comment by merlimat** [comment]
> @merlimat thanks for your contribution. For this PR, do we need to update docs?

No docs updates, this is just an internal refactoring.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 1 above)
Comment (merlimat) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java]: Done

**Review by eolivelli — APPROVED** [review]
LGTM 
I see that the patch contains the fix on MLTransactionLog

we can merge that patch and then rebase this one

**Comment by merlimat** [comment]
> I see that the patch contains the fix on MLTransactionLog

Yes, I've added here to make sure all the other tests were fine. I'll rebase on master once the other one is merged

**Pull request merged**